File: Chapter09/Challenge07/multiplying_binaries.cpp
#include <iostream>
using namespace std;

int multiplying_binaries(int p, int q) {
	int result = 0;

	while(q) {
		if (q & 1) {
			result += p;
		}
		p = p << 1;
		q = q >> 1;
	}
	
	return result;
}

void main() {
	int p, q;
	cin >> p >> q;
	cout << multiplying_binaries(p, q) <<endl;
}

File: Chapter09/Challenge07/multiplying_binaries.js
const p = 124;
const q = 29;

const is_zero = (n) => {
	return n === 0;
}

const bitsMultiply = (p, q) => {
	let result = 0;

	while (!is_zero(q)) {
		if (q & 1 !== 0)
			result += p;
		
		p = p << 1;
		q = q >> 1;
	}

	return result;
};

console.log(bitsMultiply(p, q));